File: src/main/java/com/lumiyaviewer/lumiya/render/avatar/AvatarRunningAnimation.java
package com.lumiyaviewer.lumiya.render.avatar;

import com.lumiyaviewer.lumiya.slproto.types.LLQuaternion;
import com.lumiyaviewer.lumiya.slproto.types.LLVector3;
import javax.annotation.Nonnull;

class AvatarRunningAnimation implements Comparable<AvatarRunningAnimation> {
    @Nonnull
    private final AnimationJointSet jointSet;
    @Nonnull
    private final AvatarRunningSequence sequence;

    AvatarRunningAnimation(@Nonnull AvatarRunningSequence avatarRunningSequence, @Nonnull AnimationJointSet animationJointSet) {
        this.sequence = avatarRunningSequence;
        this.jointSet = animationJointSet;
    }

    void animate(AvatarSkeleton avatarSkeleton, float[] fArr, float[] fArr2, LLQuaternion[] lLQuaternionArr, LLVector3[] lLVector3Arr) {
        this.jointSet.animate(avatarSkeleton, this.sequence, fArr, fArr2, lLQuaternionArr, lLVector3Arr);
    }

    public int compareTo(@Nonnull AvatarRunningAnimation avatarRunningAnimation) {
        int priority = avatarRunningAnimation.jointSet.getPriority() - this.jointSet.getPriority();
        if (priority != 0) {
            return priority;
        }
        priority = avatarRunningAnimation.sequence.sequenceID - this.sequence.sequenceID;
        return priority != 0 ? priority : 0;
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/avatar/AvatarRunningAnimation.kt
package com.lumiyaviewer.lumiya.render.avatar

import com.lumiyaviewer.lumiya.slproto.types.LLQuaternion
import com.lumiyaviewer.lumiya.slproto.types.LLVector3

internal class AvatarRunningAnimation(
    private val sequence: AvatarRunningSequence,
    private val jointSet: AnimationJointSet,
) : Comparable<AvatarRunningAnimation> {
    fun animate(
        avatarSkeleton: AvatarSkeleton,
        rotations: FloatArray,
        positions: FloatArray,
        quaternions: Array<LLQuaternion>,
        vectors: Array<LLVector3>,
    ) {
        jointSet.animate(avatarSkeleton, sequence, rotations, positions, quaternions, vectors)
    }

    override fun compareTo(other: AvatarRunningAnimation): Int {
        val priorityDiff = other.jointSet.priority - this.jointSet.priority
        if (priorityDiff != 0) {
            return priorityDiff
        }

        val sequenceDiff = other.sequence.sequenceID - this.sequence.sequenceID
        return if (sequenceDiff != 0) sequenceDiff else 0
    }
}
